test/composite/chain.jl

```diff
@@ -6,7 +6,7 @@ using Test, YaoBase, YaoBlockTree, YaoArrayRegister
         kron(2, 1=>phase(0.1)))
 
     @test g isa ChainBlock{2, ComplexF64} # default type
-    @test g.blocks == [kron(2, X, Y), kron(2, 1=>phase(0.1))]
+    @test g.blocks == (kron(2, X, Y), kron(2, 1=>phase(0.1)))
```

Comment:
structure transform should be immutable, it should only be able to generate an expression instead of mutating an expression, or the compiler can assume nothing on the expression, making runtime expression mutable is highly inefficient.

tag block should not be replaced, or this might cause previous inference un-proofable, (mutable runtime expression), note the block tree is a runtime expression instead of compile time expression, thus, unlike `Expr`, using `Vector` is not efficient.

Like what's in deep learning frameworks, you should create a new chain if you want to change something: https://github.com/FluxML/Flux.jl/blob/master/src/layers/basic.jl#L19

This is less convenient in syntax, but we can always polish more syntax sugars.

I'll provide a dynamic chain, and also a tuple based chain.  But dynamic chain is recommended.

Reply:
FYI: Flux was using mutable chain before as well:

https://github.com/FluxML/Flux.jl/commit/3d41dca33871ee1a25e443bfe47d2e5f291091b9